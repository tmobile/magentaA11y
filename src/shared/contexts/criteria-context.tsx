import {
  ContentTab,
  SavedCriteria,
} from 'components/content-display/markdown-content/markdown-content.types';
import React, { createContext, ReactNode, useContext } from 'react';

const LOCAL_STORAGE_KEY: string = 'savedCriteria';

interface CriteriaContextType {
  savedCriteria: SavedCriteria[];
  saveCriteria: (criteria: SavedCriteria) => void;
  removeCriteria: (id: string) => void;
  clearCriteria: (label: string) => void;
}

const CriteriaContext = createContext<CriteriaContextType | undefined>(
  undefined
);

const saveToLocalStorage = (criteria: SavedCriteria[]) => {
  try {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(criteria));
  } catch (error) {
    console.error('Failed to save criteria to localStorage:', error);
  }
};

const getInitialSavedCriteria = (): SavedCriteria[] => {
  try {
    const storedCriteria = localStorage.getItem(LOCAL_STORAGE_KEY);
    return storedCriteria ? JSON.parse(storedCriteria) : [];
  } catch (error) {
    console.error('Failed to load criteria from localStorage:', error);
    return [];
  }
};

const useSavedCriteria = () => {
  const [savedCriteria, setSavedCriteria] = React.useState<SavedCriteria[]>(
    getInitialSavedCriteria
  );

  const saveCriteria = React.useCallback((criteria: SavedCriteria) => {
    setSavedCriteria((prev) => {
      if (
        !prev.some(
          (item) =>
            item.label === criteria.label && item.content === criteria.content
        )
      ) {
        const updatedCriteria = [...prev, criteria];
        saveToLocalStorage(updatedCriteria);
        return updatedCriteria;
      }
      return prev;
    });
  }, []);

  const removeCriteria = React.useCallback((id: string) => {
    setSavedCriteria((prev) => {
      const updatedCriteria = prev.filter((item) => item.id !== id);
      saveToLocalStorage(updatedCriteria);
      return updatedCriteria;
    });
  }, []);

  const clearCriteria = React.useCallback((criteria: string) => {
    setSavedCriteria((prev) => {
      const updatedCriteria = prev.filter(
        (item: ContentTab) => item.tab !== criteria
      );
      saveToLocalStorage(updatedCriteria);
      return updatedCriteria;
    });
  }, []);

  return {
    savedCriteria,
    saveCriteria,
    removeCriteria,
    clearCriteria,
  };
};

export const CriteriaProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const criteria = useSavedCriteria();

  return (
    <CriteriaContext.Provider value={criteria}>
      {children}
    </CriteriaContext.Provider>
  );
};

export const useCriteria = () => {
  const context = useContext(CriteriaContext);
  if (!context) {
    throw new Error('useCriteria must be used within a CriteriaProvider');
  }
  return context;
};
