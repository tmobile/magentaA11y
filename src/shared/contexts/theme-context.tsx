import React, { createContext, useContext, useEffect, useState } from "react";

export enum Theme {
  LIGHT = "light",
  DARK = "dark",
}

// Types for the context
interface ThemeContextType {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}

// Default context value
const ThemeContext = createContext<ThemeContextType>({
  theme: Theme.LIGHT,
  setTheme: () => {},
});

// Provider component
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [theme, setTheme] = useState<Theme>(() => {
    const savedTheme = localStorage.getItem("theme");
    return savedTheme === Theme.DARK || savedTheme === Theme.LIGHT
      ? (savedTheme as Theme)
      : window.matchMedia("(prefers-color-scheme: dark)").matches
      ? Theme.DARK
      : Theme.LIGHT;
  });

  // Apply the theme to the document
  useEffect(() => {
    document.documentElement.setAttribute("data-theme", theme);
  }, [theme]);

  // Listen for system theme changes
  useEffect(() => {
    const systemThemeChangeListener = (e: MediaQueryListEvent) => {
      const systemTheme = e.matches ? Theme.DARK : Theme.LIGHT;
      const savedTheme = localStorage.getItem("theme");

      // Only update theme if user hasn't manually overridden
      if (!savedTheme) {
        setTheme(systemTheme);
      }
    };

    const darkModeMediaQuery = window.matchMedia(
      "(prefers-color-scheme: dark)"
    );
    darkModeMediaQuery.addEventListener("change", systemThemeChangeListener);

    // Cleanup listener on unmount
    return () => {
      darkModeMediaQuery.removeEventListener(
        "change",
        systemThemeChangeListener
      );
    };
  }, []);

  // Update localStorage only when the user explicitly changes the theme
  const handleSetTheme = (newTheme: Theme) => {
    setTheme(newTheme);
    localStorage.setItem("theme", newTheme);
  };

  return (
    <ThemeContext.Provider value={{ theme, setTheme: handleSetTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

// Custom hook for accessing the theme context
export const useTheme = () => useContext(ThemeContext);
