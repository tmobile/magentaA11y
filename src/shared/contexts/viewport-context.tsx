import { createContext, useContext, useEffect, useState } from "react";

interface ViewportContextProps {
  isMobile: boolean;
  isLargeTablet: boolean;
}

const ViewportContext = createContext<ViewportContextProps | undefined>(
  undefined
);

export const ViewportProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [isMobile, setIsMobile] = useState(false);
  const [isLargeTablet, setIsLargeTablet] = useState(false);

  useEffect(() => {
    const mobileMediaQuery = window.matchMedia("(max-width: 768px)");
    const tabletMediaQuery = window.matchMedia("(max-width: 936px)");

    setIsMobile(mobileMediaQuery.matches);
    setIsLargeTablet(tabletMediaQuery.matches);

    const handleMobileResize = () => setIsMobile(mobileMediaQuery.matches);
    const handleTableteResize = () => setIsLargeTablet(tabletMediaQuery.matches);
    mobileMediaQuery.addEventListener("change", handleMobileResize);
    tabletMediaQuery.addEventListener("change", handleTableteResize);

    return () => {
      mobileMediaQuery.removeEventListener("change", handleMobileResize);
      tabletMediaQuery.removeEventListener("change", handleTableteResize);
    };
  }, []);

  return (
    <ViewportContext.Provider value={{ isMobile, isLargeTablet }}>
      {children}
    </ViewportContext.Provider>
  );
};

export const useViewport = (): ViewportContextProps => {
  const context = useContext(ViewportContext);
  if (!context) {
    throw new Error("useViewport must be used within a ViewportProvider");
  }
  return context;
};
