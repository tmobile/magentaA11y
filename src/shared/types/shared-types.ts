export enum DocumentationCategory {
  WEB = 'web',
  NATIVE = 'native',
  HOW_TO_TEST = 'how-to-test',
}
