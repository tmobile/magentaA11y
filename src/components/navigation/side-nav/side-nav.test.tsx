import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { ViewportProvider } from 'shared/contexts/viewport-context';
import SideNav from './side-nav';
import { DocumentationCategory } from 'shared/types/shared-types';
import contentData from 'shared/content.json';
import userEvent from '@testing-library/user-event';

const renderWithProviders = (ui: React.ReactElement, isMobile = false) => {
  window.matchMedia = jest.fn().mockImplementation((query) => ({
    matches: isMobile && query === '(max-width: 768px)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }));

  return render(
    <ViewportProvider>
      <MemoryRouter
        future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
        {ui}
      </MemoryRouter>
    </ViewportProvider>
  );
};

describe('SideNav Component - Snapshot Test', () => {
  test('matches the snapshot in desktop view', () => {
    const { asFragment } = renderWithProviders(
      <SideNav documentation={DocumentationCategory.WEB} />,
      false
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('matches the snapshot in mobile view', () => {
    const { asFragment } = renderWithProviders(
      <SideNav documentation={DocumentationCategory.WEB} />,
      true
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('SideNav Component - Rendering', () => {
  test('renders the SideNav component', () => {
    renderWithProviders(<SideNav documentation={DocumentationCategory.WEB} />);
    expect(screen.getByRole('list')).toBeInTheDocument();
  });

  test('contains the correct class name', () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    const sideNav = screen.getByTestId('test-side-nav');
    expect(sideNav).toHaveClass('MagentaA11y__side-nav-container');
  });

  test('renders the title correctly', () => {
    renderWithProviders(<SideNav documentation={DocumentationCategory.WEB} />);
    expect(
      screen.getByRole('heading', { name: 'Criteria' })
    ).toBeInTheDocument();
  });

  test('renders navigation list', () => {
    renderWithProviders(<SideNav documentation={DocumentationCategory.WEB} />);
    expect(screen.getByRole('list')).toBeInTheDocument();
  });

  test('renders in a dialog element in mobile view', () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />,
      true
    );
    expect(screen.getByTestId('test-side-nav-dialog')).toBeInTheDocument();
  });

  test('renders correct nav item labels', () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    const navItems = contentData[DocumentationCategory.WEB].map(
      (item) => item.label
    );

    navItems.forEach((label) => {
      expect(screen.getByText(label)).toBeInTheDocument();
    });
  });

  test('each nav link has a valid href', async () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    for (const item of contentData[DocumentationCategory.WEB]) {
      const accordionButton = screen.getByRole('button', { name: item.label });
      await userEvent.click(accordionButton);

      const links = await screen.findAllByRole('link', { name: 'Overview' });

      const correctLink = links.find((link) =>
        link.getAttribute('href')?.includes(item.name)
      );

      expect(correctLink).toBeInTheDocument();
      expect(correctLink).toHaveAttribute(
        'href',
        expect.stringContaining(item.name)
      );
    }
  });
});

describe('SideNav Component - Interaction Tests', () => {
  test('clicking an accordion expands and collapses it', async () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    const accordionButton = screen.getByRole('button', {
      name: 'Component',
    });

    // Initially collapsed
    expect(accordionButton).toHaveAttribute('aria-expanded', 'false');

    // Click to expand
    await userEvent.click(accordionButton);
    expect(accordionButton).toHaveAttribute('aria-expanded', 'true');

    // Click again to collapse
    await userEvent.click(accordionButton);
    expect(accordionButton).toHaveAttribute('aria-expanded', 'false');
  });

  test('clicking a navigation link inside an accordion closes the menu (Mobile Only)', async () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    const accordionButton = screen.getByRole('button', {
      name: 'Page Level',
    });

    // Open the accordion
    await userEvent.click(accordionButton);

    // Click the first navigation link inside the expanded section
    const firstLink = await screen.findByRole('link', { name: 'Overview' });
    await userEvent.click(firstLink);

    // Expect the mobile side nav dialog to be closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  test('keyboard navigation with Tab key cycles through elements correctly', async () => {
    renderWithProviders(
      <SideNav
        documentation={DocumentationCategory.WEB}
        testId="test-side-nav"
      />
    );

    const accordionButton = screen.getByRole('button', {
      name: 'Component',
    });
    await userEvent.click(accordionButton);

    const links = await screen.findAllByRole('link');

    // Focus should move through each link
    for (const link of links) {
      await userEvent.tab();
      expect(link).toHaveFocus();
    }
  });
});
