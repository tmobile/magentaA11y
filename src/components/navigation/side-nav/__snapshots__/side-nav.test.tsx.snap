// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav Component - Snapshot Test matches the snapshot in desktop view 1`] = `
<DocumentFragment>
  <div
    class="MagentaA11y__side-nav-container"
  >
    <div
      class="MagentaA11y__side-nav"
    >
      <div
        class="MagentaA11y__side-nav--title-wrapper"
      >
        <h2
          class="MagentaA11y__side-nav--title"
          id="side-nav-title"
        >
          Criteria
        </h2>
      </div>
      <ul
        class="MagentaA11y__side-nav--list"
      >
        <li
          class="MagentaA11y__side-nav--item"
        >
          <div
            class="MagentaA11y-accordion"
          >
            <h3
              class="MagentaA11y-accordion__heading"
            >
              <button
                aria-expanded="false"
                class="MagentaA11y-accordion__headline"
                type="button"
              >
                <span
                  class="MagentaA11y-accordion__headline--text"
                >
                  Component
                </span>
              </button>
            </h3>
          </div>
        </li>
        <li
          class="MagentaA11y__side-nav--item"
        >
          <div
            class="MagentaA11y-accordion"
          >
            <h3
              class="MagentaA11y-accordion__heading"
            >
              <button
                aria-expanded="false"
                class="MagentaA11y-accordion__headline"
                type="button"
              >
                <span
                  class="MagentaA11y-accordion__headline--text"
                >
                  Page Level
                </span>
              </button>
            </h3>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SideNav Component - Snapshot Test matches the snapshot in mobile view 1`] = `
<DocumentFragment>
  <dialog
    aria-labelledby="side-nav-title"
    aria-modal="true"
    id="side-nav-dialog"
  >
    <div
      class="MagentaA11y__side-nav-container"
    >
      <div
        class="MagentaA11y__side-nav"
      >
        <div
          class="MagentaA11y__side-nav--title-wrapper"
        >
          <h2
            class="MagentaA11y__side-nav--title"
            id="side-nav-title"
          >
            Criteria
          </h2>
          <button
            aria-label="close"
            class="MagentaA11y-icon-button"
            data-size="large"
            id="close-side-nav-btn"
            type="button"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.40002 18.6538L5.34619 17.6L10.9462 12L5.34619 6.40002L6.40002 5.34619L12 10.9462L17.6 5.34619L18.6538 6.40002L13.0538 12L18.6538 17.6L17.6 18.6538L12 13.0538L6.40002 18.6538Z"
              />
            </svg>
          </button>
        </div>
        <ul
          class="MagentaA11y__side-nav--list"
        >
          <li
            class="MagentaA11y__side-nav--item"
          >
            <div
              class="MagentaA11y-accordion"
            >
              <h3
                class="MagentaA11y-accordion__heading"
              >
                <button
                  aria-expanded="false"
                  class="MagentaA11y-accordion__headline"
                  type="button"
                >
                  <span
                    class="MagentaA11y-accordion__headline--text"
                  >
                    Component
                  </span>
                </button>
              </h3>
            </div>
          </li>
          <li
            class="MagentaA11y__side-nav--item"
          >
            <div
              class="MagentaA11y-accordion"
            >
              <h3
                class="MagentaA11y-accordion__heading"
              >
                <button
                  aria-expanded="false"
                  class="MagentaA11y-accordion__headline"
                  type="button"
                >
                  <span
                    class="MagentaA11y-accordion__headline--text"
                  >
                    Page Level
                  </span>
                </button>
              </h3>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </dialog>
</DocumentFragment>
`;
